IdentityCore/src/throttling/metadata/MSIDThrottlingMetaDataCache.m

```diff
@@ -0,0 +1,102 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.  
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDThrottlingMetaDataCache.h"
+#import "MSIDCacheKey.h"
+#import "MSIDRequestContext.h"
+#import "MSIDThrottlingMetaData.h"
+#import "MSIDExtendedTokenCacheDataSource.h"
+#import "MSIDConstants.h"
+#import "MSIDCacheItemJsonSerializer.h"
+#import "MSIDKeychainTokenCache.h"
+#import "NSDate+MSIDExtensions.h"
+
+@implementation MSIDThrottlingMetaDataCache
+id<MSIDExtendedTokenCacheDataSource> dataSource;
+
++ (MSIDThrottlingMetaData *)getThrottlingMetadataWithAccessGroup:(NSString *)accessGroup
+                                                         Context:(id<MSIDRequestContext>)context
+                                                           error:(NSError*__nullable*__nullable)error
+{
+    MSIDThrottlingMetaData *result = nil;
+    id<MSIDExtendedTokenCacheDataSource> datasource = [MSIDThrottlingMetaDataCache dataSource:accessGroup error:error];
+    if (!datasource) return nil;
+    NSArray *jsonObjects = [datasource jsonObjectsWithKey:[MSIDThrottlingMetaDataCache throttlingMetadataCacheKey]
+                                               serializer:[MSIDCacheItemJsonSerializer new]
+                                                  context:context error:error];
+    if (![jsonObjects count])
+    {
+        return nil;
+    }
+    result = [[MSIDThrottlingMetaData alloc] initWithJSONDictionary:[jsonObjects[0] jsonDictionary] error:error];
+    return result;
+}
+
++ (BOOL)updateLastRefreshTimeWithAccessGroup:(NSString *)accessGroup
+                                     Context:(id<MSIDRequestContext>)context
+                                       error:(NSError*__nullable*__nullable)error
+{
+    NSMutableDictionary *dict = [NSMutableDictionary new];
+    dict[@"last_refresh_time"] = [[NSDate new] msidDateToTimestamp];
+
+    MSIDJsonObject *metadata = [[MSIDJsonObject alloc] initWithJSONDictionary:dict error:error];
+    if (!metadata)
+    {
+        return NO;
+    }
+    
+    id<MSIDExtendedTokenCacheDataSource> datasource = [MSIDThrottlingMetaDataCache dataSource:accessGroup error:error];
+    if (!datasource)
+    {
+        return NO;
+    }
+
+    return [datasource saveJsonObject:metadata
+                           serializer:[MSIDCacheItemJsonSerializer new]
+                                  key:[MSIDThrottlingMetaDataCache throttlingMetadataCacheKey]
+                              context:context
+                                error:error];
+}
+
++ (id<MSIDExtendedTokenCacheDataSource>)dataSource:(NSString *)accessGroup error:(NSError * _Nullable __autoreleasing * _Nullable)error
+{
+    return [[MSIDKeychainTokenCache alloc] initWithGroup:accessGroup error:error];
```

Comment:
Can we avoid initializing keychaintokencache every time?

Reply:
One option to not initializing it every time is to cache the object, for e.g <Key, Obj> is <accessGroup, keychaintokencache object>
However I don't see we cache that object anywhere, so still hesitate to do that.
Or Am I missing something here?

Reply:
I think for other cases we pass it around, and don't create a new one every time.